examples/ecs/component_hooks.rs

```diff
@@ -0,0 +1,81 @@
+//! This examples illustrates the different ways you can employ component lifecycle hooks
+
+use bevy::prelude::*;
+use std::collections::HashMap;
+
+#[derive(Component, Debug)]
+struct MyComponent(KeyCode);
+
+#[derive(Resource, Default, Debug, Deref, DerefMut)]
+struct MyComponentIndex(HashMap<KeyCode, Entity>);
+
+#[derive(Event)]
+struct MyEvent;
+
+fn main() {
+    App::new()
+        .add_plugins(DefaultPlugins)
+        .add_systems(Startup, setup)
+        .add_systems(Update, trigger_hooks)
+        .init_resource::<MyComponentIndex>()
+        .add_event::<MyEvent>()
+        .run();
+}
+
+fn setup(world: &mut World) {
+    // In order to register component hooks it must be the first time you register the component
+    // This is to prevent users from overwriting hooks that may be used internally in foreign crates
+    world
+        .register_component::<MyComponent>()
```

Comment:
If the `register_component` is only called in `setup` doesn't this mean that adding any system that reference `MyComponent` (e.g. in a `Query`) in `main` will make this fail (since they call `init_component` when they are added, which will make `register_component` fail when the system is run later on)?

Reply:
You're right, the real restraint I want to express is that there can be no archetypes containing the component when hooks are registered. I will look into expressing that more directly as you are right that is a bit of a disability footgun.